File: mobile/src/main/java/com/siliconlabs/bledemo/views/ServiceItemContainer.java
package com.siliconlabs.bledemo.views;

import android.content.Context;

import androidx.cardview.widget.CardView;

import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.utils.BLEUtils;

import java.util.HashMap;
import java.util.Map;

import butterknife.ButterKnife;
import butterknife.InjectView;

public class ServiceItemContainer extends LinearLayout {
    private int ANIMATION_DURATION_FOR_EXPAND_AND_COLLAPSE = 333;

    @InjectView(R.id.service_info_card_view)
    public CardView serviceInfoCardView;
    @InjectView(R.id.btn_expand_to_show_characteristics)
    public Button btnExpandToShowCharacteristics;
    @InjectView(R.id.service_title)
    public TextView serviceTitleTextView;
    @InjectView(R.id.image_view_edit_service)
    public ImageView serviceEditNameImageView;
    @InjectView(R.id.linear_layout_edit_service_name)
    public LinearLayout serviceEditNameLinearLayout;
    @InjectView(R.id.sevice_uuid)
    public TextView serviceUuidTextView;
    @InjectView(R.id.container_of_characteristics_for_service)
    public LinearLayout groupOfCharacteristicsForService;
    @InjectView(R.id.last_item_divider)
    public LinearLayout lastItemDivider;

    private Map<String, BLEUtils.Notifications> characteristicNotificationStates;

    public ServiceItemContainer(Context context) {
        super(context);
        init(context);
    }

    public ServiceItemContainer(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public ServiceItemContainer(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(final Context context) {
        LayoutInflater inflater = LayoutInflater.from(context);
        inflater.inflate(R.layout.list_item_debug_mode_service, this);

        ButterKnife.inject(this);

        btnExpandToShowCharacteristics.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (groupOfCharacteristicsForService.getVisibility() == View.VISIBLE) {
                    btnExpandToShowCharacteristics.setText(getResources().getString(R.string.More_Info));
//                    animateCaretCollapse(context, serviceExpansionCaret);\
                    groupOfCharacteristicsForService.setVisibility(GONE);
                } else {
                    btnExpandToShowCharacteristics.setText(getResources().getString(R.string.Less_Info));
                    animateCharacteristicExpansion();
                }
            }
        });

        serviceInfoCardView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (groupOfCharacteristicsForService.getVisibility() == View.VISIBLE) {
                    btnExpandToShowCharacteristics.setText(getResources().getString(R.string.More_Info));
//                    animateCaretCollapse(context, serviceExpansionCaret);\
                    groupOfCharacteristicsForService.setVisibility(GONE);
                } else {
                    btnExpandToShowCharacteristics.setText(getResources().getString(R.string.Less_Info));
                    animateCharacteristicExpansion();
                }
            }
        });

        characteristicNotificationStates = new HashMap<>();
    }

    private void animateCharacteristicExpansion() {
        final View characteristicsExpansion = this.groupOfCharacteristicsForService;

        characteristicsExpansion.measure(FlowLayout.LayoutParams.MATCH_PARENT, FlowLayout.LayoutParams.WRAP_CONTENT);
        final int targetHeight = characteristicsExpansion.getMeasuredHeight();

        // Older versions of android (pre API 21) cancel animations for views with a height of 0.
        characteristicsExpansion.getLayoutParams().height = 1;
        characteristicsExpansion.setVisibility(View.VISIBLE);
        Animation animation = new Animation() {
            @Override
            protected void applyTransformation(float interpolatedTime, Transformation t) {
                characteristicsExpansion.getLayoutParams().height = interpolatedTime == 1 ? FlowLayout.LayoutParams.WRAP_CONTENT : (int) (targetHeight * interpolatedTime);
                characteristicsExpansion.requestLayout();
            }

            @Override
            public boolean willChangeBounds() {
                return true;
            }
        };

        animation.setDuration(ANIMATION_DURATION_FOR_EXPAND_AND_COLLAPSE);
        characteristicsExpansion.startAnimation(animation);
    }

    public void setCharacteristicNotificationState(String characteristicName, BLEUtils.Notifications state) {
        characteristicNotificationStates.put(characteristicName, state);
    }

    public BLEUtils.Notifications getCharacteristicNotificationState(String characteristicName) {
        return characteristicNotificationStates.get(characteristicName);
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Views/ServiceItemContainer.kt
package com.siliconlabs.bledemo.views

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.cardview.widget.CardView
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.utils.BLEUtils.Notifications

class ServiceItemContainer(context: Context) : LinearLayout(context) {
    private var characteristicNotificationStates: MutableMap<String, Notifications> = HashMap()

    private val btnExpandToShowCharacteristics: Button
    val cvServiceInfo: CardView
    val tvServiceTitle: TextView
    val ivEditServiceName: ImageView
    val llServiceEditName: LinearLayout
    val tvServiceUuid: TextView
    val llGroupOfCharacteristicsForService: LinearLayout
    val llLastItemDivider: LinearLayout

    companion object {
        private const val ANIMATION_DURATION_FOR_EXPAND_AND_COLLAPSE = 333
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.list_item_debug_mode_service, this)

        cvServiceInfo = findViewById(R.id.service_info_card_view)
        btnExpandToShowCharacteristics = findViewById(R.id.btn_expand_to_show_characteristics)
        tvServiceTitle = findViewById(R.id.service_title)
        ivEditServiceName = findViewById(R.id.image_view_edit_service)
        llServiceEditName = findViewById(R.id.linear_layout_edit_service_name)
        tvServiceUuid = findViewById(R.id.sevice_uuid)
        llGroupOfCharacteristicsForService = findViewById(R.id.container_of_characteristics_for_service)
        llLastItemDivider = findViewById(R.id.last_item_divider)

        btnExpandToShowCharacteristics.setOnClickListener {
            if (llGroupOfCharacteristicsForService.visibility == View.VISIBLE) {
                btnExpandToShowCharacteristics.text = resources.getString(R.string.More_Info)
                llGroupOfCharacteristicsForService.visibility = View.GONE
            } else {
                btnExpandToShowCharacteristics.text = resources.getString(R.string.Less_Info)
                animateCharacteristicExpansion()
            }
        }
        cvServiceInfo.setOnClickListener {
            if (llGroupOfCharacteristicsForService.visibility == View.VISIBLE) {
                btnExpandToShowCharacteristics.text = resources.getString(R.string.More_Info)
                llGroupOfCharacteristicsForService.visibility = View.GONE
            } else {
                btnExpandToShowCharacteristics.text = resources.getString(R.string.Less_Info)
                animateCharacteristicExpansion()
            }
        }
    }

    private fun animateCharacteristicExpansion() {
        val characteristicsExpansion: View? = llGroupOfCharacteristicsForService
        characteristicsExpansion?.measure(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.WRAP_CONTENT)
        val targetHeight = characteristicsExpansion?.measuredHeight

        // Older versions of android (pre API 21) cancel animations for views with a height of 0.
        characteristicsExpansion?.layoutParams?.height = 1
        characteristicsExpansion?.visibility = View.VISIBLE
        val animation: Animation = object : Animation() {
            override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                characteristicsExpansion?.layoutParams?.height = if (interpolatedTime == 1f) ViewGroup.LayoutParams.WRAP_CONTENT else (targetHeight!! * interpolatedTime).toInt()
                characteristicsExpansion?.requestLayout()
            }

            override fun willChangeBounds(): Boolean {
                return true
            }
        }
        animation.duration = ANIMATION_DURATION_FOR_EXPAND_AND_COLLAPSE.toLong()
        characteristicsExpansion?.startAnimation(animation)
    }

    fun setCharacteristicNotificationState(characteristicName: String, state: Notifications) {
        characteristicNotificationStates[characteristicName] = state
    }

}